witness/golang/cmd/witness/internal/witness/witness.go

```diff
@@ -117,48 +115,93 @@ func (w *Witness) GetCheckpoint(logID string) ([]byte, error) {
 	}
 	cosigned, err := note.Sign(n, w.Signer)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("couldn't sign checkpoint: %v", err)
 	}
 	return cosigned, nil
 }
 
 // Update updates the latest checkpoint if chkptRaw is consistent with the current
-// latest one for this log (according to latestSize).   It returns the latest
-// checkpoint size before the update was applied (or just what was fetched if the
-// update was unsuccessful).
-func (w *Witness) Update(ctx context.Context, logID string, latestSize uint64, chkptRaw []byte, proof [][]byte) (uint64, error) {
+// latest one for this log.   It returns the latest checkpoint size before the
+// update was applied (or just what was fetched if the update was unsuccessful).
+func (w *Witness) Update(ctx context.Context, logID string, chkptRaw []byte, proof [][]byte) (uint64, error) {
+	// If we don't witness this log then no point in going further.
+	logInfo, ok := w.Logs[logID]
+	if !ok {
+		return 0, fmt.Errorf("no information for log %q", logID)
+	}
 	// Check the signatures on the raw checkpoint and parse it
 	// into the log.Checkpoint format.
 	chkpt, err := w.parse(chkptRaw, logID)
 	if err != nil {
 		return 0, err
 	}
+	c := Chkpt{Size: chkpt.Size, Raw: chkptRaw}
 	// Get the latest one for the log because we don't want consistency proofs
-	// with respect to older checkpoints.
-	latest, err := w.db.GetLatest(logID)
+	// with respect to older checkpoints.  Bind this all in a transaction to
+	// avoid race conditions when updating the database.
+	tx, err := w.db.db.BeginTx(ctx, nil)
```

Comment:
The necessity to use `db.db` reinforces my point above that `Database` type seems unnecessary. There is no clear boundary between `Database` in `Witness` anyway, so it could just all be inside the `Witness`.

Reply:
Agreed, in line with the earlier comment.